map/traffic_manager.cpp

```diff
@@ -215,3 +325,86 @@ void TrafficManager::CheckCacheSize()
     }
   }
 }
+
+bool TrafficManager::IsEnabled() const
+{
+  return m_state != TrafficState::Disabled;
+}
+
+bool TrafficManager::IsInvalidState() const
+{
+  return m_state == TrafficState::NetworkError;
+}
+
+void TrafficManager::UpdateState()
+{
+  if (!IsEnabled() || IsInvalidState())
+    return;
+
+  auto const currentTime = steady_clock::now();
+  auto maxPassedTime = steady_clock::duration::zero();
+
+  bool waiting = false;
+  bool networkError = false;
+  bool expiredApp = false;
+  bool expiredMwm = false;
+  bool noData = false;
+
+  for (auto const & mwmId : m_activeMwms)
+  {
+    auto it = m_mwmCache.find(mwmId);
+    ASSERT(it != m_mwmCache.end(), ());
+
+    if (it->second.m_isWaitingForResponse)
+    {
+      waiting = true;
+    }
+    else
+    {
+      expiredApp |= it->second.m_lastAvailability == traffic::TrafficInfo::Availability::ExpiredApp;
+      expiredMwm |= it->second.m_lastAvailability == traffic::TrafficInfo::Availability::ExpiredMwm;
+      noData |= it->second.m_lastAvailability == traffic::TrafficInfo::Availability::NoData;
+    }
+
+    if (it->second.m_isLoaded)
+    {
+      auto const timeSinceLastResponse = currentTime - it->second.m_lastResponseTime;
+      if (timeSinceLastResponse > maxPassedTime)
+        maxPassedTime = timeSinceLastResponse;
+    }
+    else if (it->second.m_retriesCount >= kMaxRetriesCount)
+    {
+      networkError = true;
+    }
+  }
+
+  if (networkError || maxPassedTime >= kNetworkErrorTimeout)
+    ChangeState(TrafficState::NetworkError);
+  else if (waiting)
+    ChangeState(TrafficState::WaitingData);
+  else if (expiredApp)
+    ChangeState(TrafficState::ExpiredApp);
+  else if (expiredMwm)
+    ChangeState(TrafficState::ExpiredData);
+  else if (noData)
+    ChangeState(TrafficState::NoData);
+  else if (maxPassedTime >= kOutdatedDataTimeout)
+    ChangeState(TrafficState::Outdated);
+  else
+    ChangeState(TrafficState::Enabled);
+}
+
+void TrafficManager::ChangeState(TrafficState newState)
+{
+  if (m_state == newState)
+    return;
+
+  m_state = newState;
+
+  GetPlatform().RunOnGuiThread([this]()
```

Comment:
I guess it's important to capture state by value here, otherwise the callback may be invoked on changed m_state. Even better is to capture newState.